File: 2-add-two-numbers/2-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        
        total = self.readList(l1) + self.readList(l2)
        head = ListNode()
        cur = head
        
        while total:
            cur.next = ListNode(total % 10)
            cur = cur.next
            total //= 10
            
        return head.next if head.next else head
        
    
    def readList(self, head: Optional[ListNode]) -> int:
        
        node = head
        result = 0
        power = 0
        while node:
            result += node.val * 10**power
            node = node.next 
            power += 1
            
        return result

File: 2-add-two-numbers/2-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        return recursiveHelper(l1, l2, 0);
    }
    
    private ListNode recursiveHelper(ListNode l1, ListNode l2, int remainder) {
        if (l1 != null && l2 != null) {
            remainder += l1.val + l2.val;
            l1.val = remainder % 10;
            l1.next = recursiveHelper(l1.next, l2.next, remainder/10);
            return l1;
        } else if (l2 != null) {
            return recursiveHelper(l2, l1, remainder);
        } else if (l1 != null) {
            if (remainder != 0) {
                remainder += l1.val;
                l1.val = remainder % 10;
                l1.next = recursiveHelper(l1.next, l2, remainder/10);
            }
            
            return l1;
        } else if (remainder == 0) {
            return null;
        }
        return new ListNode(remainder);
    }
}